tests/user_metadata_test.go

```diff
@@ -0,0 +1,130 @@
+		Namespace: namespace,
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: workflowID,
+			RunId:      runID,
+		},
+	})
+}
+
+func prepareTestUserMetadata() *sdkpb.UserMetadata {
+	return &sdkpb.UserMetadata{
+		Summary: &commonpb.Payload{
+			Metadata: map[string][]byte{"test_summary_key": []byte(`test_summary_val`)},
+			Data:     []byte(`Test summary Data`),
+		},
+		Details: &commonpb.Payload{
+			Metadata: map[string][]byte{"test_details_key": []byte(`test_details_val`)},
+			Data:     []byte(`Test Details Data`),
+		},
+	}
+}
+
+// TestUserMetadataStartWorkflowExecution verifies that the user metadata passed in StartWorkflowExecution request
+// is correctly recorded and returned in the describe workflow response.
+func (s *FunctionalSuite) TestUserMetadataStartWorkflowExecution() {
+	id := "functional-user-metadata-StartWorkflowExecution"
+	metadata := prepareTestUserMetadata()
+	request := &workflowservice.StartWorkflowExecutionRequest{
+		RequestId:    uuid.New(),
+		Namespace:    s.namespace,
+		WorkflowId:   id,
+		WorkflowType: &commonpb.WorkflowType{Name: "functional-user-metadata-workflow-type"},
+		TaskQueue:    &taskqueuepb.TaskQueue{Name: "functional-user-metadata-taskqueue", Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		UserMetadata: metadata,
+	}
+
+	we, err := s.engine.StartWorkflowExecution(NewContext(), request)
+	s.NoError(err)
+
+	// Verify that the UserMetadata associated with the start event is returned in the describe response.
+	describeInfo, err := getDescribeWorkflowExecutionInfo(s.engine, s.namespace, id, we.RunId)
+	s.NoError(err)
+	s.NotNil(describeInfo.ExecutionConfig.UserMetadata)
+	s.EqualExportedValues(describeInfo.ExecutionConfig.UserMetadata, metadata)
+}
+
+// TestUserMetadataSignalWithStartWorkflowExecution verifies that the user metadata attached to SignalWithStartWorkflowExecution request
+// is correctly recorded and returned in the describe workflow response.
+func (s *FunctionalSuite) TestUserMetadataSignalWithStartWorkflowExecution() {
+	id := "functional-user-metadata-SignalWithStartWorkflowExecution"
+	metadata := prepareTestUserMetadata()
+	request := &workflowservice.SignalWithStartWorkflowExecutionRequest{
+		RequestId:    uuid.New(),
+		Namespace:    s.namespace,
+		WorkflowId:   id,
+		WorkflowType: &commonpb.WorkflowType{Name: "functional-user-metadata-workflow-type"},
+		TaskQueue:    &taskqueuepb.TaskQueue{Name: "functional-user-metadata-taskqueue", Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		SignalName:   "TEST-SIGNAL",
+		UserMetadata: metadata,
+	}
+
+	we, err := s.engine.SignalWithStartWorkflowExecution(NewContext(), request)
+	s.NoError(err)
+
+	// Verify that the UserMetadata associated with the start event is returned in the describe response.
+	describeInfo, err := getDescribeWorkflowExecutionInfo(s.engine, s.namespace, id, we.RunId)
+	s.NoError(err)
+	s.NotNil(describeInfo.ExecutionConfig.UserMetadata)
+	s.EqualExportedValues(describeInfo.ExecutionConfig.UserMetadata, metadata)
```

Comment:
I didn't know about that assertion! I've seen others use `s.ProtoEqual` to do the same thing.

Note that most (not all) assertions in testify have the expected value _first_.

Reply:
expected before actual is definitely the convention, testify will label the two values that way when it prints a diff so it's quite confusing if they're swapped

Reply:
and I also didn't know about EqualExportedValues. maybe we could actually replace our uses of ProtoEqual with that? @tdeebswihart 
ProtoEqual is our own code and it'd be nice to not have to maintain it